hunting/__main__.py

```diff
@@ -0,0 +1,212 @@
+# 2.0.
+
+import textwrap
+from pathlib import Path
+
+import click
+from tabulate import tabulate
+
+from detection_rules.misc import parse_user_config
+
+from .definitions import HUNTING_DIR
+from .markdown import MarkdownGenerator
+from .run import QueryRunner
+from .search import QueryIndex
+from .utils import get_hunt_path, load_toml, update_index_yml, filter_elasticsearch_params
+
+
+@click.group()
+def hunting():
+    """Commands for managing hunting queries and converting TOML to Markdown."""
+    pass
+
+
+@hunting.command('generate-markdown')
+@click.argument('path', required=False)
+def generate_markdown(path):
+    """Convert TOML hunting queries to Markdown format."""
+    markdown_generator = MarkdownGenerator(HUNTING_DIR)
+
+    if path:
+        path = Path(path)
+        if path.is_file() and path.suffix == '.toml':
+            click.echo(f"Generating Markdown for single file: {path}")
+            markdown_generator.process_file(path)
+        elif (HUNTING_DIR / path).is_dir():
+            click.echo(f"Generating Markdown for folder: {path}")
+            markdown_generator.process_folder(path)
+        else:
+            click.echo(f"Invalid path: {path}. It should be a valid TOML file or a folder.")
+    else:
+        click.echo("Generating Markdown for all files.")
+        markdown_generator.process_all_files()
+
+    # After processing, update the index
+    markdown_generator.update_index_md()
+
+
+@hunting.command('refresh-index')
+def refresh_index():
+    """Refresh the index.yml file from TOML files and then refresh the index.md file."""
+    click.echo("Refreshing the index.yml and index.md files.")
+    update_index_yml(HUNTING_DIR)
+    markdown_generator = MarkdownGenerator(HUNTING_DIR)
+    markdown_generator.update_index_md()
+    click.echo("Index refresh complete.")
+
+
+@hunting.command('search')
+@click.option('--tactic', type=str, default=None, help="Search by MITRE tactic ID (e.g., TA0001)")
+@click.option('--technique', type=str, default=None, help="Search by MITRE technique ID (e.g., T1078)")
+@click.option('--sub-technique', type=str, default=None, help="Search by MITRE sub-technique ID (e.g., T1078.001)")
+@click.option('--data-source', type=str, default=None, help="Filter by data_source like 'aws', 'macos', or 'linux'")
+def search_queries(tactic: str, technique: str, sub_technique: str, data_source: str):
+    """Search for queries based on MITRE tactic, technique, sub-technique, or data_source."""
+
+    if not any([tactic, technique, sub_technique, data_source]):
+        raise click.UsageError("""Please provide at least one filter (tactic, technique, sub-technique,
+                               or data_source) to search queries.""")
+
+    click.echo("Searching for queries based on provided filters...")
+
+    # Filter out None values from the MITRE filter tuple
+    mitre_filters = tuple(filter(None, (tactic, technique, sub_technique)))
```

Comment:
Good question. MITRE is something consistent to filter on because it's required but also adds a constraint to what can be searched based on a popular framework.

We can add text-based searching from the notes, also tags is another option if we implemented them. I assume people may want to search for anything `powershell` or `cloud` or `Lumma` too. That is a good point but will likely require us introduce tags in a later PR (since it will require us to update ALL hunting queries. For now, we can do a `keyword` option that searches in description, notes and name.